msal/src/main/java/com/microsoft/identity/client/statemachine/results/SignInResult.kt

```diff
@@ -0,0 +1,173 @@
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.client.statemachine.results
+
+import com.microsoft.identity.client.statemachine.BrowserRequiredError
+import com.microsoft.identity.client.statemachine.Error
+import com.microsoft.identity.client.statemachine.IncorrectCodeError
+import com.microsoft.identity.client.statemachine.PasswordIncorrectError
+import com.microsoft.identity.client.statemachine.UserNotFoundError
+import com.microsoft.identity.client.statemachine.states.AccountResult
+import com.microsoft.identity.client.statemachine.states.SignInCodeRequiredState
+import com.microsoft.identity.client.statemachine.states.SignInPasswordRequiredState
+
+/**
+ * Sign in result, produced by
+ * [com.microsoft.identity.client.INativeAuthPublicClientApplication.signIn]
+ */
+sealed interface SignInResult : Result {
+
+    /**
+     * Complete Result, which indicates the sign in flow is complete,
+     * i.e. tokens are retrieved for the provided scopes (if any).
+     *
+     * @param resultValue an [com.microsoft.identity.client.statemachine.states.AccountResult] object containing account information and account related methods.
+     */
+    class Complete(override val resultValue: AccountResult) :
+        Result.CompleteResult(resultValue = resultValue),
+        SignInResult,
+        SignInSubmitCodeResult,
+        SignInUsingPasswordResult,
+        SignInSubmitPasswordResult
+
+    /**
+     * BrowserRequired ErrorResult, which indicates that the server requires more/different authentication mechanisms than the client is configured to provide.
+     * The flow should be restarted with a browser, by calling [com.microsoft.identity.client.IPublicClientApplication.acquireToken]
+     *
+     * @param error [com.microsoft.identity.client.statemachine.BrowserRequiredError].
+     */
+    class BrowserRequired(override val error: BrowserRequiredError) :
+        Result.ErrorResult(error = error),
+        SignInResult,
+        SignInUsingPasswordResult,
+        SignInSubmitCodeResult,
+        SignInResendCodeResult,
+        SignInSubmitPasswordResult
+
+    /**
+     * UnexpectedError ErrorResult is a general error wrapper which indicates an unexpected error occurred during the flow.
+     * If this occurs, the flow should be restarted.
+     *
+     * @param error [com.microsoft.identity.client.statemachine.Error].
+     */
+    class UnexpectedError(override val error: Error) :
+        Result.ErrorResult(error = error),
```

Comment:
Mentioned in another place about not having so many unnecessary duplicated classes, and instead one result or error class that can hold different error codes.

Reply:
I will discuss it with the team but won't be able to make changes as part of this PR as it will be significant refactoring.